dataflow/gemma-flex-template/README.md

```diff
@@ -0,0 +1,179 @@
+
+        Example Usage::
+
+          pcoll | RunInference(GemmaPytorchHandler())
+
+        Args:
+          model_variant: The Gemma model name.
+          checkpoint_path: the path to a local copy of gemma model weights.
+          tokenizer_path: the path to a local copy of the gemma tokenizer
+          device: optional. the device to run inference on. can be either
+            'cpu' or 'gpu', defaults to cpu. 
+        """
+        model_config = get_config_for_2b(
+        ) if "2b" in model_variant else get_config_for_7b()
+        model_config.tokenizer = tokenizer_path
+        model_config.quant = 'quant' in model_variant
+        model_config.tokenizer = tokenizer_path
+
+        self._model_config = model_config
+        self._checkpoint_path = checkpoint_path
+        if device == 'GPU':
+            logging.info("Device is set to CUDA")
+            self._device = torch.device('cuda')
+        else:
+            logging.info("Device is set to CPU")
+            self._device = torch.device('cpu')
+        self._env_vars = {}
+
+    def load_model(self) -> GemmaForCausalLM:
+        """Loads and initializes a model for processing."""
+        torch.set_default_dtype(self._model_config.get_dtype())
+        model = GemmaForCausalLM(self._model_config)
+        model.load_weights(self._checkpoint_path)
+        model = model.to(self._device).eval()
+        return model
+```
+
+### Formatting DoFn
+
+The output from a keyed model handler is a tuple of the form `(key, PredictionResult)`. To format that output into a string before sending it to the answer Pub/Sub topic, use an extra `DoFn`.
+
+```py
+class FormatOutput(beam.DoFn):
+  def process(self, element, *args, **kwargs):
+    yield "Key : {key}, Input: {input}, Output: {output}".format(key=element[0], input=element[1].example, output=element[1].inference)
+```
+
+## Build the Flex Template
+Run the following code from the directory to build the Dataflow flex template.
+
+- Replace `$TEMPLATE_FILE` with a Google Cloud Storage path written as a JSON file.
+- Set `SDK_CONTAINER_IMAGE` to the name of the Docker image created previously.
+- `$PROJECT` is the Google Cloud project that you created previously. 
+
+```sh
+gcloud dataflow flex-template build $TEMPLATE_FILE \
+  --image $SDK_CONTAINER_IMAGE \
+  --sdk-language "PYTHON" \
+  --metadata-file metadata.json \
+  --project $PROJECT
+```
+
+## Start the pipeline
+To start the Dataflow streaming job, run the following code from the directory. Replace `$TEMPLATE_FILE`, `$REGION`, `$GCS_BUCKET`, `$INPUT_SUBSCRIPTION`, `$OUTPUT_TOPIC`, `$SDK_CONTAINER_IMAGE`, and `$PROJECT` with the Google Cloud project resources you created previously. It might take as much as 15 minutes for the worker to start up and to begin accepting messages from the input Pub/Sub topic. 
+
+```sh
+gcloud dataflow flex-template run "gemma-flex-`date +%Y%m%d-%H%M%S`" \
+  --template-file-gcs-location $TEMPLATE_FILE \
+  --region $REGION \
```

Comment:
I could not use us-central1 due to capacity constraints, but trying asia-southeast1 caused knockon effects, because this invocation isn't specifying a temporary location, and a default bucket was created for me ("default_gcs_bucket_name" method in [apache_beam.io.gcp.gcsio](https://beam.apache.org/releases/pydoc/2.29.0/_modules/apache_beam/io/gcp/gcsio.html#GcsIO) according to the log files) that was invalid because the bucket name was too long due to the length of the region name. I was able to use asia-east1 to get around this issue, but I think the higher level solution is ensuring that this invocation specifies a `--temp-location` (unsure if this needs to be a parameter of gcloud or an additional `--parameters`.

Reply:
I also think this is the cause of the [failing tests](https://github.com/GoogleCloudPlatform/python-docs-samples/pull/11881#issuecomment-2189254462). In the e2e testing, you're defaulting to us-central1, which I was having contention issues with, and checking the testing project logs, it's also having issues (I can't isolate if the logs I'm seeing are from this test or the other gemma sample, but I think it's the same issue.)

Reply:
the temp location change is relatively easy, just an extra flag that can share the same bucket as the staging location